app/controllers/admin/vouchers_controller.rb

```diff
@@ -22,6 +22,11 @@ def create
     rescue ActiveRecord::SubclassNotFound
       @voucher.errors.add(:type)
       render_error
+    rescue ActiveRecord::RecordNotUnique
+      # Rails unique validation doesn't work with soft deleted object, so we rescue the database
+      # exception  to display a nice message to the user
+      @voucher.errors.add(:code, :taken)
+      render_error
```

Comment:
It seems like something that the `paranoia` gem could handle for us. But I can't find any mention of that, so this is good 👍